rfcs/geometric-primitives.md

```diff
@@ -0,0 +1,349 @@
+# Feature Name: `geometric-primitives`
+
+## Summary
+
+Lightweight geometric primitive types for use across bevy engine crates, and as interoperability types for external libraries.
+
+## Motivation
+
+This would provide a standard way to model primitives across the bevy ecosystem to prevent ecosystem fragmentation amongst plugins.
+
+## User-Facing Explanation
+
+Geometric primitives are lightweight representations of geometry that describe the type of geometry as well as fully defined dimensions. These primitives are *not* meshes, but the underlying precise mathematical definition. For example, a circle is:
+
+```rust
+pub struct Circle2d {
+  origin: Point2d,
+  radius: f32,
+}
+```
+
+Geometric primitives have a defined shape, size, position, and orientation. Position and orientation are **not** defined using bevy's `Transform` components. This is because these are fundamental geometric primitives that must be usable and comparable as-is
+
+`bevy_geom` provides two main modules, `geom_2d` and `geom_3d`. Recall that the purpose of this crate is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `geom_2d::Line2d` and `geom_3d::Line`. Note that the 2d version of a line is lighter weight, and is only defined in 2d. 3d geometry (or 2d with depth which is 3d) is assumed to be the default for most cases. The names of the types were chosen with this in mind, to guide you toward using Line instead of Line2d for example, unless you know why you are making this choice.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+
+### 3D Geometry Types
+
+These types are fully defined in 3d space.
+
+```rust
+/// Point in 3D space
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+
+// Plane in 3d space defined by a point on the plane, and the normal direction of the plane
+struct Plane {
+  point: Point,
+  normal: Direction,
+}
+
+/// Unbounded line in 3D space with direction
+struct Ray { 
+  point: Point, 
+  normal: Direction,
+}
+
+/// Line in 3D space bounded by two points
+struct Line { 
+  start: Point, 
+  end: Point,
+}
+
+struct Triangle([Point; 3]);
+
+struct Quad([Point; 4]);
+
+struct Sphere{
+  origin: Point,
+  radius: f32,
+}
+
+struct Cylinder {
+  /// The bottom center of the cylinder.
+  origin: Point,
+  radius: f32,
+  /// The extent of the cylinder is the vector that extends from the origin to the opposite face of the cylinder.
+  extent: Vec3,
+}
+
+struct Cone{
+  /// Origin of the cone located at the center of its base.
+  origin: Point,
+  /// The extent of the cone is the vector that extends from the origin to the tip of the cone.
+  extent: Vec3,
+  base_radius: f32,
+}
+
+struct Torus{
```

Comment:
Out of curiosity, why is the Torus useful enough to warrant inclusion in the basic primitives?

Reply:
It's a common primitive, especially for constructing things like rotation handles on gizmos. 😉 

I've been struggling to explain why I've also been having this doubt about including the torus, and I think it has to do with their use in collision and bounding. Unlike the other primitives, the torus isn't convex. Maybe there needs to be a distinction between convex volumes/areas, which are what benefit from this fully-defined lightweight representation, and 2d/3d primitives that are useful for mesh generation.

Reply:
If that distinction exists, a trait may be a good way to capture it.

Reply:
I agree. I'm trying to figure out how to express this in traits without making everything overcomplicated.